lib/alg.ml

```diff
@@ -63,51 +64,80 @@ module Strict_Ternary : Alg_t = struct
     {
       s = State.create 1;
       q = Pieotree.create topology;
-      z = scheduling_transaction;
+      z_in = scheduling_transaction;
+      z_out = fun x -> fun _ -> x;
     }
 
   let simulate sim_length pkts =
     Control.simulate sim_length 0.001 poprate pkts control
 end
 
-module RRobin_Ternary : Alg_t = struct
-  let scheduling_transaction s pkt =
-    let time = Packet.time pkt in
-    let flow = Packet.find_flow pkt in
-    let var_last_finish = Printf.sprintf "%s_last_finish" (Flow.to_string flow) in
-    (* We will use this variable to read/write to state. *)
-    let rank_for_root =
-      if State.isdefined var_last_finish s then
-        max (Time.to_float time) (State.lookup var_last_finish s)
-      else Time.to_float time
-    in
-    let s' =
-      State.rebind var_last_finish (rank_for_root +. (100.0 /. 0.33)) s
-    in
-    let rank_for_root = Rank.create rank_for_root time in
-    let int_for_root =
-      (* Put flow A into leaf 0, flow B into leaf 1, and flow C into leaf 2. *)
-      match flow with
+let rrobin n = (
+  module struct
+    (* n-flow Round-Robin *)
+    let _ = assert (n <= 7 && n > 0)
+
+    let who_skip pop turn = 
+      let rec who_skip_aux t acc = 
+        if t = pop then 
+          acc
+        else
+          who_skip_aux ((t + 1) mod n) (t :: acc)
+      in
+      who_skip_aux turn []
+
+    let pkt_to_int pkt =
+      match Packet.find_flow pkt with
       | A -> 0
       | B -> 1
-      | C -> 2
-      | n -> failwith Printf.(sprintf "Don't know how to route flow %s." (Flow.to_string n))
-    in
-    (* The ranks are as calculated above. *)
-    ([ (int_for_root, rank_for_root); (0, Rank.create 0.0 time) ], s', Time.epoch)
-
-  let topology = Topo.one_level_ternary
-
-  let control : Control.t =
-    {
-      s = State.create 3;
-      q = Pieotree.create topology;
-      z = scheduling_transaction;
-    }
-
-  let simulate sim_length pkts =
-    Control.simulate sim_length 0.001 poprate pkts control
-end
+      | C -> 2 
+      | D -> 3 
+      | E -> 4 
+      | F -> 5 
+      | G -> 6
+
+    let z_in s pkt = 
+      let time = Packet.time pkt in
+      let int_for_root = pkt_to_int pkt in
+      let r_i = "r_" ^ (string_of_int int_for_root) in
+      let rank_for_root = State.lookup r_i s in
+      let s' = State.rebind r_i (rank_for_root +. (float_of_int n)) s in
+      let rank_for_root = Rank.create rank_for_root time in
+      ([ (int_for_root, rank_for_root); (0, Rank.create 0.0 time) ], s', Time.epoch)
+
+    let z_out s pkt = 
+      let turn = State.lookup "turn" s in
+      let turn' = ((pkt_to_int pkt) + 1) mod n |> float_of_int in 
+      let s' = State.rebind "turn" turn' s in
+      let skipped = who_skip (pkt_to_int pkt) (int_of_float turn) in
+      let f s i = 
+        let r_i = "r_" ^ (string_of_int i) in
+        State.rebind r_i (State.lookup r_i s +. (float_of_int n)) s
+      in
+      List.fold_left f s' skipped
```

Comment:
Super cool! Would you like to add a sketch of the algorithm in here as a comment? Either that, or you could leave it be for a sec (with a TODO), later add a note on the Calyx docs website that explains the algorithm, and then update this code to point to that documentation. Your call!

Reply:
I've moved [this github comment](https://github.com/cucapra/schedulers-in-ocaml/pull/2#issuecomment-2268607486) into comments in `rrobin` with [18780f0](https://github.com/cucapra/schedulers-in-ocaml/pull/2/commits/18780f03749e6831baf22592fc248b5bc8fe7787)